src/gradualizer_lib.erl

```diff
@@ -236,6 +237,34 @@ get_ast_children(Node) ->
     [_Tag, _Anno | Children] = tuple_to_list(Node),
     Children.
 
+%% @doc Sort a list of types according to their relative position in the subtyping hierarchy.
+-spec type_sort([typechecker:type()], tenv()) -> [typechecker:type()].
+type_sort(Tys, TEnv) ->
+    lists:sort(fun (Ty1, Ty2) ->
+                       subtype_less_or_eq(Ty1, Ty2, TEnv)
+               end, Tys).
+
+%% @doc Compare two types according to their relative position in the subtyping hierarchy.
+%% Type `Ty1' is less or equal than another type `Ty2' if it's its subtype,
+%% but `Ty2' is not `Ty1''s subtype.
+%% To ensure total ordering, if both types are subtypes of each other or they're not subtypes at all
+%% the total order over Erlang terms is used for comparison.
+-spec subtype_less_or_eq(typechecker:type(), typechecker:type(), tenv()) -> boolean().
+subtype_less_or_eq(Ty1, Ty2, TEnv) ->
+    case {typechecker:subtype(Ty1, Ty2, TEnv), typechecker:subtype(Ty2, Ty1, TEnv)} of
+        {{true, _}, {true, _}} ->
+            %% Cannot determine order in the type hierarchy.
+            %% Compare using Erlang term total order.
+            Ty1 =< Ty2;
```

Comment:
I'm worried about reordering the keys in a map in this way.

If we ignore the unknown (or dynamic) type, it's probably fine to do this,
because if two types are subtypes of each other, it means they're equal.
That's not the case with gradual types though. We have `1..10 :: any()` but also
`any() :: 1..10`.

We don't know at compile time what this particular instance of `any()`
represents. It may be `1..5` or `integer()` or anything else; the programmer
might know but the type checker doesn't.

So, for example, A = `#{1..10 => atom(), any() => pid()` is not equivalent to B
= `#{any() => pid(), 1..10 => atom()`. The expression `#{5 => self()}` is of
type B but it's not of type A.

If we want to reorder the keys, we first have to make them disjoint.

I've seen in a paper, don't remember where but perhaps in
https://www.irif.fr/~gc/papers/icfp17.pdf (Gradual Typing with Union and
Intersection Types, Castagna & Lanvin), a method to do this. It involves
intersection and negation types. By example, it goes like this:

1. `#{1..10 => atom(), any() => pid()` -- Start here and let's reorder the keys
2. `#{1..10 => atom(), any() \ 1..10 => pid()` -- This is equivalent to the
above type, since the leftmost key takes precedence. The type `any() \ 1..10`
means "unknown, but at least we know it can't be anything in the range 1..10`.
3. #{any() \ 1..10 => pid(), 1..10 => atom()` -- Now the keys are disjoint, so they can be safely reordered.

The "set minus" is short-hand for "and not", so `any() & ~1..10` with another
notation.

To be able to do things properly like GLB for map types and to handle function
specs with multiple clauses (to make them disjoint in the same way), I think we
need union, intersection and negation types, but so far we have only
intersection types.

Reply:
Thanks for the examples and pointers. I've started reading the referenced paper, but following further ones it refers to seems like a rabbit hole - it's going to take a while before I grok them fully.

> I think we need union, intersection and negation types, but so far we have only intersection types.

I think we only have union types, don't we? Having started reading the above paper it seems to me that intersection and negation types would be those enabling us to express `any() \ 1..10`.

That being said, I'd like to try this approach, but given I can only make small steps in that direction, I'm afraid it's going to take a while :|

---

However, to salvage as much as possible from this PR:
- I've completely removed sorting map field associations when normalising map type or computing the GLB
- I've short-circuited `glb()` on maps to return `none()` if keys of any of the map params overlap - in other words, we return a GLB in more situations than before, but still decline to compute it if negation types are required to do it properly

This means the meaningful changes this PR carries are:
- map type inference of map creation expressions
- `subtype` (or actually `compat_ty`) logic changes which follow the wiki
- handling of map unions in `update_map_type`
- map normalisation, so they're easier to handle

To sum up, https://github.com/erszcz/Gradualizer/tree/map-tests is a fork of the Gradualizer master branch with tests from this PR added on top - it fails on 15 test cases. This PR passes all of these 15 cases.

If you think it's reasonable to introduce such an incremental improvement, instead of a complete solution with intersection types, I'll clean up this PR and prepare it for review. In such a case I think it's best to start the review from the tests side - https://github.com/erszcz/Gradualizer/tree/map-tests - that is by checking if changes I made there are valid.

Otherwise, I think we can close this PR while I (or someone else) work on a proper implementation with intersection types.

Reply:
Good summary @erszcz. I think this is great work and want to see it merged!

And yes of course I meant union types. Sorry for that, if it confused anyone.

@josefs has stated before that he thinks the type representation in that paper is unnecessarily complex and that our representation of types using the erl_parse's abstract format is good enough.

I've considered rewriting the whole thing to using that representation from the paper, but that would be a lot of work, as you say. That paper doesn't solve all our problems anyway, for example polymorphism with type variables isn't solved AFAIK.

Another rewrite we could do first is to merge `type_check_expr` with `type_check_expr_in` into one that both takes a type to check against and returns a type, instead of one function returning a type and the other checking against a given type and not returning a type. Similarly for `type_check_pat(_in)` and others. That may simplify things a bit and get rid of code duplication.

Reply:
Whew, it's taken a while, but I've finally cleaned the code up. The commit history is still quite messy, but I guess it's not a problem if we consider a squash merge - please let me know if we want to keep it, though, I'll clean it up in such a case.

Reply:
Nice! I think hou can squash the commits and update the commit message to a summary of what was done finally.

I will look at the code tomorrow (or the next days).